sagemcom_api/client.py

```diff
@@ -190,57 +253,15 @@ async def __api_request_async(self, actions, priority=False):
         }
 
         form_data = {"req": json.dumps(payload, separators=(",", ":"))}
-
-        async with self.session.post(api_host, data=form_data) as response:
-            if response.status == 400:
-                result = await response.text()
-                raise BadRequestException(result)
-
-            if response.status != 200:
-                result = await response.text()
-                raise UnknownException(result)
-
-            if response.status == 200:
-                result = await response.json()
-                error = self.__get_response_error(result)
-
-                # No errors
-                if (
-                    error["description"] == XMO_REQUEST_NO_ERR
-                    or error["description"] == "Ok"  # NOQA: W503
-                ):
-                    return result
-
-                # Error in one of the actions
-                if error["description"] == XMO_REQUEST_ACTION_ERR:
-                    # pylint:disable=fixme
-                    # TODO How to support multiple actions + error handling?
-                    actions = result["reply"]["actions"]
-                    for action in actions:
-                        action_error = action["error"]
-                        action_error_description = action_error["description"]
-
-                        if action_error_description == XMO_NO_ERR:
-                            continue
-
-                        if action_error_description == XMO_AUTHENTICATION_ERR:
-                            raise AuthenticationException(action_error)
-
-                        if action_error_description == XMO_ACCESS_RESTRICTION_ERR:
-                            raise AccessRestrictionException(action_error)
-
-                        if action_error_description == XMO_NON_WRITABLE_PARAMETER_ERR:
-                            raise NonWritableParameterException(action_error)
-
-                        if action_error_description == XMO_UNKNOWN_PATH_ERR:
-                            raise UnknownPathException(action_error)
-
-                        if action_error_description == XMO_MAX_SESSION_COUNT_ERR:
-                            raise MaximumSessionCountException(action_error)
-
-                        raise UnknownException(action_error)
-
-                return result
+        try:
+            result = await self.__post(api_host, form_data)
+            return result
+        except (
+            ClientConnectorError,
+            ClientOSError,
+            ServerDisconnectedError,
+        ) as exception:
+            raise ConnectionError(str(exception)) from exception
```

Comment:
Why would we want to raise a more general `ConnectionError` exception, instead of the underlying exception?

See https://docs.aiohttp.org/en/v3.8.2/client_reference.html#hierarchy-of-exceptions. You can actually already use [ClientError](https://docs.aiohttp.org/en/v3.8.2/client_reference.html#aiohttp.ClientError) to catch broader exceptions.

Reply:
There are two main reasons in my mind:

1. To hide the internals of the API. If we ever change the underlying aiohttp library, aiohttp specific exceptions like ClientConnectorError or ClientError will be a source of issues later.
2. Raising a single type of built-in exception (ConnectionError) simplifies the error handling logic. Consumers of the API don't need to worry about the specific connection errors that might occur; they can simply catch ConnectionError and handle it appropriately.